File: 678. Valid Parenthesis String.py
class Solution:
    def checkValidString(self, s):
        """
        :type s: str
        :rtype: bool
        """
        n = len(s)
        dp = [True]
        dp[0] = True
        dp2 = []
        for i, c in enumerate(s, 1):
            if c == '(':
                dp2 = [False] + dp                
            elif c == ')':
                dp2 = dp[1:] + [False, False]
            else:
                x = [False] + dp
                y = dp[1:] + [False, False]
                z = dp + [False]
                dp2 = [x[i] or y[i] or z[i] for i in range(i+1)]
            dp = dp2
        return dp[0]

File: 678. Valid Parenthesis String.go
func checkValidString(s string) bool {
	left := 0
	stack := []rune{}
	for _, c := range s {
		if c == '(' || c == '*' {
			if c == '(' {
				left++
			}
			stack = append(stack, c)
		} else {
			if len(stack) == 0 {
				return false
			}
			if left == 0 {
				// pop * as '('
				stack = stack[:len(stack)-1]
			} else {
				// find and pop '('
				left--
				for i := len(stack) - 1; i >= 0; i-- {
					if stack[i] == '(' {
						for j := i; j < len(stack)-1; j++ {
							stack[j] = stack[j+1]
						}
						stack = stack[:len(stack)-1]
						break
					}
				}
			}
		}
	}

	if left == 0 {
		return true
	}

	// check if there are enough * to replace ')'
	star := 0
	for i := len(stack) - 1; i >= 0; i-- {
		if stack[i] == '*' {
			star++
		} else if star > 0 {
			star--
		} else {
			return false
		}
	}
	return true
}
